language/compiler/ir_to_bytecode/syntax/src/syntax.rs

```diff
@@ -1 +1,1798 @@
+    -> Result<Var_, ParseError<usize, Token<'input>, failure::Error>>
+{
+    let v_start_loc = tokens.token.0;
+    let var = parse_var(tokens)?;
+    let v_end_loc = tokens.token.0;
+    Ok(Spanned{span: Span::new(ByteIndex(v_start_loc as u32),
+                               ByteIndex(v_end_loc as u32)),
+               value: var })
+}
+
+// Field: Field = {
+//     <n:Name> =>? parse_field(n),
+// };
+
+fn parse_field_<'input, 'builder>(tokens: &mut __Matcher<'input, 'builder>)
+    -> Result<Field_, ParseError<usize, Token<'input>, failure::Error>>
+{
+    let start_loc = tokens.token.0;
+    let f = parse_field(parse_name(tokens)?)?;
+    let end_loc = tokens.token.0;
+    Ok(Spanned{span: Span::new(ByteIndex(start_loc as u32),
+                               ByteIndex(end_loc as u32)),
+               value: f })
+}
+
+// CopyableVal: CopyableVal = {
+//     AccountAddress => CopyableVal::Address(<>),
+//     "true" => CopyableVal::Bool(true),
+//     "false" => CopyableVal::Bool(false),
+//     <i: U64> => CopyableVal::U64(i),
+//     <buf: ByteArray> => CopyableVal::ByteArray(buf),
+// }
+
+fn parse_copyable_val_<'input, 'builder>(tokens: &mut __Matcher<'input, 'builder>)
+    -> Result<CopyableVal_, ParseError<usize, Token<'input>, failure::Error>>
+{
+    let start_loc = tokens.token.0;
+    let val = match (tokens.token.1).0 {
+        Tok::AccountAddressValue => {
+            let addr = parse_account_address(tokens)?;
+            CopyableVal::Address(addr)
+        },
+        Tok::True => {
+            tokens.advance()?;
+            CopyableVal::Bool(true)
+        },
+        Tok::False => {
+            tokens.advance()?;
+            CopyableVal::Bool(false)
+        },
+        Tok::U64Value => {
+            let i = u64::from_str((tokens.token.1).1).unwrap();
+            tokens.advance()?;
+            CopyableVal::U64(i)
+        },
+        Tok::ByteArrayValue => {
+            let s = (tokens.token.1).1;
+            let buf = ByteArray::new(hex::decode(&s[2..s.len()-1]).unwrap_or_else(|_| panic!("The
+string {:?} is not a valid hex-encoded byte array", s)));
+            tokens.advance()?;
+            CopyableVal::ByteArray(buf)
+        },
+        _ => { return Err(ParseError::InvalidToken { location: tokens.token.0 })},
+    };
+    let end_loc = tokens.token.0;
+    Ok(Spanned{span: Span::new(ByteIndex(start_loc as u32),
+                               ByteIndex(end_loc as u32)),
+               value: val })
+}
+
+// Exp = BinopExp = CmpOpExp
+// CmpOpExp = (OrExp CmpOp OrExp) || OrExp
+// OrExp = (AndExp OrOp AndExp) || AndExp
+// AndExp = (XorExp AndOp XorExp) || XorExp
+// XorExp = (BinOrExp XorOp BinOrExp) || BinOrExp
+// BinOrExp = (BinAndExp BinOrOp BinAndExp) || BinAndExp
+// BinAndExp = (AddSubExp BinAndOp AddSubExp) || AddSubExp
+// AddSubExp = (FactorExp AddSubOp FactorExp) || FactorExp
+// FactorExp = (UnaryExp FactorOp UnaryExp) || UnaryExp
+
+fn get_precedence(token: &Tok) -> u32
```

Comment:
Might make sense to put this inside `impl Tok`? This would make it easier for someone adding a new token to see that they need to add precedence (particularly if we make the pattern matching exhaustive, which could be a good idea).

Couple of other pieces of logic below that might also benefit from this treatment; have noted each with "Move to `impl Tok`?".

Reply:
I think of tokens and their properties as part of the lexer. Binary operator precedence is more about the parser, so I would not expect that to be in the code for the lexer. I can see your point about making it easier for someone who is adding a new token for a binary operator, though. @tnowacki , do you have an opinion on that?

Regarding exhaustive pattern matching, I'm not sure what you mean. Are you suggesting that we list every token in every match expression? That wouldn't make sense to me, since most tokens are only relevant in a particular part of the grammar. Did you mean something different?

Reply:
I have no strong opinions on making it in the impl or not

And I don't think exhaustive pattern matching will really help all that much, but I could be wrong

Reply:
> Are you suggesting that we list every token in every match expression?

I did indeed mean this because that's the only way to let the compiler tell us what code to change when we add a new token. But I understand that it's a pain, particularly when there's a single enum of tokens.

Reply:
It's not so much that it is a pain, but more that it isn't useful. If we add a new token, it would probably only be relevant at one or two points in the grammar, so there would be only a few places in the recursive descent that would need to check for that token.